sapio-psbt/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+};
+use bitcoin::{KeyPair, TxOut};
+use bitcoin::{Network, SchnorrSig};
+use miniscript::psbt::PsbtExt;
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+use std::error::Error;
+use std::fmt::Display;
+use std::str::FromStr;
+
+#[derive(Serialize, Deserialize)]
+#[serde(untagged)]
+pub enum PSBTApi {
+    Finished {
+        completed: bool,
+        hex: String,
+    },
+    NotFinished {
+        completed: bool,
+        psbt: String,
+        error: String,
+        errors: Vec<String>,
+    },
+}
+
+pub fn finalize_psbt_format_api(psbt: PartiallySignedTransaction) -> PSBTApi {
+    let secp = Secp256k1::new();
+    psbt.finalize(&secp)
+        .map(|tx| {
+            let hex = bitcoin::consensus::encode::serialize_hex(&tx.extract_tx());
+            PSBTApi::Finished {
+                completed: true,
+                hex,
+            }
+        })
+        .unwrap_or_else(|(psbt, errors)| {
+            let errors: Vec<_> = errors.iter().map(|e| format!("{:?}", e)).collect();
+            let encoded_psbt = base64::encode(serialize(&psbt));
+            PSBTApi::NotFinished {
+                completed: false,
+                psbt: encoded_psbt,
+                error: "Could not fully finalize psbt".into(),
+                errors,
+            }
+        })
+}
+
+pub async fn read_key_from_file(
+    file_name: &std::ffi::OsStr,
+) -> Result<ExtendedPrivKey, Box<dyn Error>> {
+    let buf = tokio::fs::read(file_name).await?;
+    Ok(ExtendedPrivKey::decode(&buf)?)
+}
+
+pub fn sign(
+    xpriv: ExtendedPrivKey,
+    mut psbt: PartiallySignedTransaction,
+    hash_ty: bitcoin::SchnorrSighashType,
+) -> Result<Vec<u8>, PSBTSigningError> {
+    sign_psbt_inplace(&xpriv, &mut psbt, &Secp256k1::new(), hash_ty)?;
+    let bytes = serialize(&psbt);
+    Ok(bytes)
+}
+pub async fn show_pubkey(input: &std::ffi::OsStr) -> Result<(), Box<dyn Error>> {
+    let buf = tokio::fs::read(input).await?;
+    let xpriv = ExtendedPrivKey::decode(&buf)?;
+    println!("{}", ExtendedPubKey::from_priv(&Secp256k1::new(), &xpriv));
+    Ok(())
+}
+pub fn new_key(network: &str, out: &std::ffi::OsStr) -> Result<(), Box<dyn Error>> {
+    let entropy: [u8; 32] = rand::thread_rng().gen();
+    let xpriv = ExtendedPrivKey::new_master(Network::from_str(network)?, &entropy)?;
+    std::fs::write(out, &xpriv.encode())?;
+    println!("{}", ExtendedPubKey::from_priv(&Secp256k1::new(), &xpriv));
+    Ok(())
+}
+
+pub fn sign_psbt_inplace(
+    xpriv: &ExtendedPrivKey,
+    psbt: &mut PartiallySignedTransaction,
+    secp: &Secp256k1<All>,
+    hash_ty: bitcoin::SchnorrSighashType,
+) -> Result<(), PSBTSigningError> {
+    sign_psbt_input_inplace(xpriv, psbt, secp, 0, hash_ty)
+}
+
+#[derive(Debug, Clone)]
+pub enum PSBTSigningError {
+    NoUTXOAtIndex(usize),
+    NoInputAtIndex(usize),
+}
+
+impl Display for PSBTSigningError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
```

Comment:
Do you want your display to be the same as your Debug?

Reply:
Well, it's an Error type so it seems OK?